PG-1455 Add random base to IVs when encrypting WAL and relation data

jeltz opened the pull request:
It is unclear if this really adds much extra security since we still always use the same pair of key and IV, but since it is easy to do we might as well stop using fixed IVs and use a unique IV for each key. I add the base to the IV using XOR. An alternative would be to do a normal addition, either on a byte array or in a `uint128_t`, but xor was easy and I think just as good. Is my logic correct there?

I also for fun spent a bit of time writing the code in a way gcc could vectorize. I did not managed to get clang to vectorize, but the generated machine code looked good enough anyway. If it had been more performance critical I would have used 128 bit intrinsics but it is not.

I like how for the SMGR IVs one of the easiest ways to write it also ended up the fastest in GCC (with `-O2`), but for GCC to vectorize the WAL IV I needed to use temporary variables instead of just writing to `iv_prefix` directly.

The two final pieces of code I used in Godbolt.

```c
#include <stdint.h>
#include <string.h>
#include <stdio.h>

typedef uint32_t BlockNumber;

static void
CalcBlockIv(BlockNumber bn, const unsigned char *base_iv, unsigned char *iv)
{
	memset(iv, 0, 16);

	iv[12] = bn >> 24;
	iv[13] = bn >> 16;
	iv[14] = bn >> 8;
	iv[15] = bn;

	for (int i = 0; i < 16; i++)
		iv[i] ^= base_iv[i];
}

int f(BlockNumber bn, const unsigned char base[16]) {
    unsigned char iv[16];

    CalcBlockIv(bn, base, iv);

    printf("%s", iv);
}
```

```c
#include <stdint.h>
#include <string.h>
#include <stdio.h>

typedef uint32_t TimeLineID;
typedef uint64_t XLogRecPtr;

static void
CalcXLogPageIVPrefix(TimeLineID tli, XLogRecPtr lsn, const unsigned char *base_iv, char *iv_prefix)
{
	/* Temporary variables to make GCC optimize it */
	char 		a[16],
				b[16];

	memset(a, 0, 16);
	a[0] = tli >> 24;
	a[1] = tli >> 16;
	a[2] = tli >> 8;
	a[3] = tli;

	memset(b, 0, 16);
	b[4] = lsn >> 56;
	b[5] = lsn >> 48;
	b[6] = lsn >> 40;
	b[7] = lsn >> 32;
	b[8] = lsn >> 24;
	b[9] = lsn >> 16;
	b[10] = lsn >> 8;
	b[11] = lsn;

	for (int i = 0; i < 16; i++)
		iv_prefix[i] = (a[i] | b[i]) ^ base_iv[i]; 

	/* Zero lowest 4 bytes */
	for (int i = 12; i < 16; i++)
		iv_prefix[i] = 0;
}

int f(TimeLineID tli, XLogRecPtr lsn, const unsigned char base[16]) {
    unsigned char iv_prefix[16];

    CalcXLogPageIVPrefix(tli, lsn, base, iv_prefix);

    printf("%s", iv_prefix);
}
```

jeltz committed:
PG-1455 Add field with random base IV to relation and WAL keys

For now we do not use this field but we plan to use it when encrypting
Wrelation data and WAL.

jeltz committed:
PG-1455 Add random base numbers to IVs for relation data

We might as well increase the entropy by adding a random base value to
the IVs used used when encrypting relations. But since for now the pair
of key + IV is generated and used together it adds little extra security
over what we already have.

We add the IV with XOR since that is a cheap and easy operation which
uases no extra collissions.

jeltz committed:
PG-1455 Add random base numbers to IVs for WAL encryption

Same as last commit but for the WAL encryption.

Rewrote the calculation in a way gcc's vectorizer likes, as verified
with Godbolt. The code generated by clang is ok and branch free but it
fails to properly vectorize both before and after.

dutow commented on contrib/pg_tde/src/access/pg_tde_xlog_encrypt.c:
```diff
@@ -356,21 +357,34 @@ tdeheap_xlog_seg_read(int fd, void *buf, size_t count, off_t offset,
+	b[10] = lsn >> 8;
+	b[11] = lsn;
+
+	for (int i = 0; i < 16; i++)
```
WAL uses CTR, the IV there should be an increasing sequence from the starting point - we should add the two numbers together.

This would work, but this with this we are even further with "standard" CTR than before. Ideally, we should do an IV calculation that would allow decryption with openssl AES-CTR algorithm. We do not verify this currently anywhere.

jeltz replied:
Yeah, that is a good point and something I thought about but was too lazy to fix. And you have conviced me now that it should be fixed. I am also veyr tempted to just hack together using CTR form OpenSSL. It should not be much work at all.

jeltz replied:
Pushed a version which uses addition.

jeltz committed:
PG-1455 Change xlog base IVs to use addition

This way they can eventually be made more similar to using OpenSSL
for encrypting and decrypting the CTR stream.

jeltz commented:
My idea if we want to make the IV generation more similar to CTX is to increment the counter in a 128-bit wrapping manner and then do:

```c
static void
CalcXLogPageIVPrefix(TimeLineID tli, XLogRecPtr lsn, const unsigned char *base_iv, char *iv_prefix)
{
	union u128cast base;
	union u128cast iv;
	unsigned	__int128 offset;

	for (int i = 0; i < 16; i++)
#ifdef WORDS_BIGENDIAN
		base.a[i] = base_iv[i];
#else
		base.a[i] = base_iv[15 - i];
#endif

	offset = (((unsigned __int128) tli) << 64) | (((unsigned __int128) lsn) >> 4);

	iv.i = base.i + offset;

	for (int i = 0; i < 16; i++)
#ifdef WORDS_BIGENDIAN
		iv_prefix[i] = iv.a[i];
#else
		iv_prefix[i] = iv.a[15 - i];
#endif
}
```
